crates/client/eth/src/state_update.rs

```diff
@@ -0,0 +1,207 @@
+use alloy::primitives::Address;
+use anyhow::Context;
+use dc_db::DeoxysBackend;
+use dc_metrics::block_metrics::BlockMetrics;
+use dp_convert::ToFelt;
+use dp_transactions::TEST_CHAIN_ID;
+use dp_utils::channel_wait_or_graceful_shutdown;
+use futures::StreamExt;
+use serde::Deserialize;
+use starknet_api::hash::StarkHash;
+use starknet_types_core::felt::Felt;
+use url::Url;
+
+use crate::client::StarknetCoreContract;
+use crate::{
+    client::EthereumClient,
+    utils::{convert_log_state_update, trim_hash},
+};
+
+#[derive(Debug, Clone, Deserialize, PartialEq)]
+pub struct L1StateUpdate {
+    pub block_number: u64,
+    pub global_root: StarkHash,
+    pub block_hash: StarkHash,
+}
+
+/// Get the last Starknet state update verified on the L1
+pub async fn get_initial_state(client: &EthereumClient) -> anyhow::Result<L1StateUpdate> {
+    let block_number = client.get_last_verified_block_number().await?;
+    let block_hash = client.get_last_verified_block_hash().await?;
+    let global_root = client.get_last_state_root().await?;
+
+    Ok(L1StateUpdate { global_root, block_number, block_hash })
+}
+
+/// Subscribes to the LogStateUpdate event from the Starknet core contract and store latest
+/// verified state
+pub async fn listen_and_update_state(
+    eth_client: &EthereumClient,
+    backend: &DeoxysBackend,
+    block_metrics: &BlockMetrics,
+    chain_id: Felt,
+) -> anyhow::Result<()> {
+    let event_filter = eth_client.l1_core_contract.event_filter::<StarknetCoreContract::LogStateUpdate>();
+
+    let mut event_stream = event_filter.watch().await.context("Failed to watch event filter")?.into_stream();
+
+    while let Some(event_result) = channel_wait_or_graceful_shutdown(event_stream.next()).await {
+        let log = event_result.context("listening for events")?;
+        let format_event: L1StateUpdate =
+            convert_log_state_update(log.0.clone()).context("formatting event into an L1StateUpdate")?;
+        update_l1(backend, format_event, block_metrics, chain_id)?;
+    }
+
+    Ok(())
+}
+
+pub fn update_l1(
+    backend: &DeoxysBackend,
+    state_update: L1StateUpdate,
+    block_metrics: &BlockMetrics,
+    chain_id: Felt,
+) -> anyhow::Result<()> {
+    // This is a provisory check to avoid updating the state with an L1StateUpdate that should not have been detected
+    //
+    // TODO: Remove this check when the L1StateUpdate is properly verified
+    if state_update.block_number > 500000u64 || chain_id == TEST_CHAIN_ID {
```

Comment:
could you replace this with `if state_update.block_number > 500000u64 && chain_id == MAIN_CHAIN_ID` so that this doesn't get triggered in app chains

Reply:
fixed